src/canvasLogic/Conference.ts

```diff
@@ -0,0 +1,224 @@
+            role: this._data.role,
+        });
+        update("update_user_data", userData);
+    }
+}
+
+interface IConferenceBroadcastingEvent {
+    type: "broadcast";
+    target: CanvasId;
+    drawing: string;
+}
+
+class ConferenceBroadcastingEvent extends ConferenceEvent {
+    protected _data: IConferenceBroadcastingEvent;
+
+    private _update: (id: CanvasId, data: string) => void;
+
+    constructor(
+        data: IConferenceBroadcastingEvent,
+        conferenceId: string,
+        update: (id: CanvasId, data: string) => void
+    ) {
+        super(data, conferenceId);
+
+        this._data = data;
+        this._update = update;
+    }
+
+    handle(): void {
+        const { target, drawing } = this._data;
+        this._update(target, drawing);
+        console.log(`Got broadcast to ${target}`);
+    }
+}
+
+class IncorrectConferenceEventTypeError extends Error {
+    constructor(message: string) {
+        super(message);
+    }
+}
+
+class IncorrectConferenceEventInterfaceError extends Error {
+    constructor(message: string) {
+        super(message);
+    }
+}
+
+type Subscriber = Required<{
+  update: (data: string) => void,
+  id: CanvasId,
+}>
+
+type Controller = Required<{
+  update: (data: IClientConferenceEvent) => void,
+}>
+
+// Conference events on client
+enum ClientConferenceEventTypes_ {
+  CreateCanvas = "create_canvas",
+}
+
+export const ClientConferenceEventTypes = Object.freeze(ClientConferenceEventTypes_);
+
+export interface IClientConferenceEvent {
+  type: ClientConferenceEventTypes_,
+  data: object,
+}
+
+class ConferenceSubscribers {
+    private _subscribers: Map<CanvasId, Subscriber> = new Map();
+
+    private _controller: Controller;
+
+    constructor (controller: Controller) {
+        this._controller = controller;
+    }
+
+    public notify(id: CanvasId, data: string): void {
+        if (!this._subscribers.has(id)) {
+            this.notifyController({
+                type: ClientConferenceEventTypes.CreateCanvas,
+                data: {id: id},
+            });
+        }
+        this._subscribers.get(id)?.update(data);
+    }
+
+    private notifyController(data: IClientConferenceEvent): void {
+        this._controller.update(data);
+    }
+
+    public subscribe(canvas: Subscriber): void {
+        this._subscribers.set(canvas.id, canvas);
+    }
+}
+
+export default class Conference {
+    private static _WEBSOCKET_URL = import.meta.env.VITE_WEBSOCKET_URL;
+
+    private _webSocket: WebSocket | null = null;
+
+    private _subscribers: ConferenceSubscribers = new ConferenceSubscribers(this);
+
+    private _controller: Controller;
+
+    private _userData: {id: string, role: number} | null = null;
+
+    constructor(controller: Controller, conferenceId: string) {
+        this.join(conferenceId);
+
+        this._controller = controller;
+    }
+
+    private join(conferenceId: string): void {
+        this._webSocket = new WebSocket(Conference._WEBSOCKET_URL + conferenceId);
```

Comment:
Does it handle connection errors anyhow?

Reply:
No, it would be just displayed error in debug mode and log in console in production